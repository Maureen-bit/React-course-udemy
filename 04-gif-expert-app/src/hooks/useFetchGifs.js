import { useEffect, useState } from "react"
import { getGifs } from "../helpers/getGIF";

export const useFetchGifts = (category ) => {

    /* Custom hooks son diferentes a functional components, el custom hook se 
    encarga de hacer el trabajo pesado, de la logica fuerte, como obtener informaciÃ³n de las Api y retornarlas al 
    functional component para que las muestre */
    const [state, setstate] = useState({
        data: [],
        loading: true
    });

    useEffect(()=>{
        getGifs(category)
            .then( images => 
                setstate({
                    data: images,
                    loading: false
                })
            )
    }, [category]);

    return state;
}