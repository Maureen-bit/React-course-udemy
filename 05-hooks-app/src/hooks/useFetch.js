import { useEffect, useRef, useState } from 'react';

export const useFetch = ( url ) => {
    /* Voy a guardar la referencia  de cuando se monta el componente, como se renderiza por primera vez
    entonces este valor será true*/
    const isMounted = useRef(true);
    const [state, setState] = useState({data: null, loading: true, error: null});

    useEffect(() => {    
      return () => {
          /* Este return se disparará cada vez que se desmonte el componente por eso cambiamos el valor de la
          referencia a false */
        isMounted.current = false;
      };
    }, []);
    
    useEffect(() => {
        setState({data: null, loading: true, error: null});
        fetch(url)
        .then(resp => resp.json())
        .then(data => {
            /* Usamos el valor de la referencia para ejecutar el setState cuando el componente
            esté montado y evitar el error de: 
            Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory 
            leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function. */
            if (isMounted.current){
                /* Si está en true es porque el componente está montado */
                setState({
                    loading: false,
                    error: null,
                    data
                })
            } else {
                console.log("SetState no se llamó")
            }
        })
    }, [url]);

    return state;
};
