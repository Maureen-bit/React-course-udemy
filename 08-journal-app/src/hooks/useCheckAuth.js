import { onAuthStateChanged } from 'firebase/auth';
import { useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { FirebaseAuth } from '../firebase/config';
import { login, logout } from '../store/auth';

export const useCheckAuth = () => {
    const { status } = useSelector(state => state.auth);
    const dispatch = useDispatch();
    
    useEffect(() => {
      /** onAuthStateChanged: función que está pendiente de los cambios en el estado de la autenticación del usuario */
      onAuthStateChanged(FirebaseAuth, async (user) => {
        if(!user){
          return dispatch(logout());
        }
  
        const { uid, email, displayName, photoURL } = user;
        dispatch(login({uid, email, displayName, photoURL}))
  
      });
  
    }, [])
  return { status }
}
