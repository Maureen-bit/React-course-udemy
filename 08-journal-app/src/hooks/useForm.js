import { useEffect, useMemo, useState } from 'react';

export const useForm = ( initialForm = {}, formValidations = {} ) => {
  
    const [ formState, setFormState ] = useState( initialForm );
    /** Cuando requerimos volver a renderizar por cambios en el estado debemos usar un hook que maneje el estado */
    const [ formValidation, setFormValidation ] = useState( {} );

    useEffect(() => {
        createValidators();
    }, [formState]);

    const isFormValid = useMemo(() =>{
        for (const formValue of Object.keys(formValidation)) {
            if( formValidation[formValue] !== null) {
                return false;
            }
        }
        return true;
    }, [formValidation]);

    const onInputChange = ({ target }) => {
        const { name, value } = target;
        setFormState({
            ...formState,
            [ name ]: value
        });
    }

    const onResetForm = () => {
        setFormState( initialForm );
    }

    const createValidators = () => {
        const formCheckedValues = {};
        for (const formField of Object.keys(formValidations)) {
            /**  Devuelve cada arreglo de cada propiedad (email, password, displayName)*/
            const [ funcionValidacion, errorMessage = 'Este valor es requerido' ] = formValidations[formField];
            /** creación de variable dinámica dentro del arreglo para cada propiedad (email, password, displayName), 
             * según resultado de la función, si es válido no devuelve nada porque no hay error pero si hay error devuelve el mensaje
             * de error.
             */
            formCheckedValues[`${ formField}Valid`] = funcionValidacion(formState[formField]) ? null : errorMessage;
        }
        setFormValidation(formCheckedValues);
    }



    return {
        ...formState,
        formState,
        onInputChange,
        onResetForm,
        ...formValidation,
        isFormValid
    }
}