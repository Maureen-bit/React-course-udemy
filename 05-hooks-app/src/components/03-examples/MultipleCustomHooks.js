import { useCounter } from '../../hooks/useCounter';
import { useFetch } from '../../hooks/useFetch';
import '../02-useEffect/effects.css';

export const MultipleCustomHooks = () => {
    const {counter, increment } = useCounter(1);
    const {loading, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);
    const { author, quote } = !!data && data[0];
    /* !null = true, !!null = false  */
 
    return (
        <>
            <h1> Breaking bad </h1>
            <hr />

            {loading  ? 
                (
                <div className='alert alert-info text-center'>
                    Loading...
                </div> 
                )
             :
                (
                    <blockquote className='blockquote text-center'>
                        <p className='mb-2'> {quote} </p>
                        <footer className='blockquote-footer' > {author} </footer>
                    </blockquote>
                )
            }

            <button className='btn btn-primary' onClick={increment}>
                Siguiente quote!
            </button>
        </>
    )
};
