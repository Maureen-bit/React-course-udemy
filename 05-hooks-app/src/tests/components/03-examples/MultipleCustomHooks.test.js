import React from 'react';
import { shallow } from 'enzyme';
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';

/* Al tratar de usar el custom hook aquí en este archivo se realizará una implementación mockeada, este mock
evita las llamadas al hook, en esta prueba no importa como está implementado el useFerch porque ya se probó */
jest.mock('../../../hooks/useFetch')
jest.mock('../../../hooks/useCounter')

describe('Multiple custom hooks tests', () => { 

    test('should render correctly', () => { 
        useCounter.mockReturnValue({
            counter: 10,
            increment: () => {}
        })
        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        })
        const wrapper = shallow( < MultipleCustomHooks />)
        expect(wrapper).toMatchSnapshot();
     })

     test('should show information', () => { 
        useCounter.mockReturnValue({
            counter: 10,
            increment: () => {}
        })
        useFetch.mockReturnValue({
            data: [{
                author: "Maureen",
                quote: "amo a Candy"
            }],
            loading: false,
            error: null
        })
        const wrapper = shallow( < MultipleCustomHooks />)
        /* esta alerta no debe existir porque ya hay información renderizada */
        expect(wrapper.find('.alert').exists()).toBe(false);
        expect(wrapper.find('p').text().trim()).toBe('amo a Candy');
        expect(wrapper.find('footer').text().trim()).toBe('Maureen');
      })
 })