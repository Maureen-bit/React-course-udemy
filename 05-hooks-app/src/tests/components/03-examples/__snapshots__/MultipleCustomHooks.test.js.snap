// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multiple custom hooks tests should render correctly 1`] = `
<Fragment>
  <h1>
     Breaking bad 
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente quote!
  </button>
</Fragment>
`;
